bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonDecoder.kt

```diff
@@ -154,15 +156,18 @@ internal open class DefaultBsonDecoder(
         }
     }
 
-    override fun decodeByte(): Byte = decodeInt().toByte()
-    override fun decodeChar(): Char = decodeString().single()
-    override fun decodeFloat(): Float = decodeDouble().toFloat()
-    override fun decodeShort(): Short = decodeInt().toShort()
-    override fun decodeBoolean(): Boolean = readOrThrow({ reader.readBoolean() }, BsonType.BOOLEAN)
-    override fun decodeDouble(): Double = readOrThrow({ reader.readDouble() }, BsonType.DOUBLE)
-    override fun decodeInt(): Int = readOrThrow({ reader.readInt32() }, BsonType.INT32)
-    override fun decodeLong(): Long = readOrThrow({ reader.readInt64() }, BsonType.INT64)
-    override fun decodeString(): String = readOrThrow({ reader.readString() }, BsonType.STRING)
+    override fun decodeByte(): Byte = NumberCodecHelper.decodeByte(reader)
+
+    override fun decodeChar(): Char = StringCodecHelper.decodeChar(reader)
+    override fun decodeFloat(): Float = NumberCodecHelper.decodeFloat(reader)
+
+    override fun decodeShort(): Short = NumberCodecHelper.decodeShort(reader)
+    override fun decodeBoolean(): Boolean = reader.readBoolean()
+
+    override fun decodeDouble(): Double = NumberCodecHelper.decodeDouble(reader)
+    override fun decodeInt(): Int = NumberCodecHelper.decodeInt(reader)
+    override fun decodeLong(): Long = NumberCodecHelper.decodeLong(reader)
+    override fun decodeString(): String = reader.readString()
```

Comment:
Lets keep the `readOrThrow({ reader.readString() }, BsonType.STRING)` so that users know if an invalid type is present for a string field.